apps/aefate/src/aefa_data.erl

```diff
@@ -0,0 +1,117 @@
+-type fate_map()     :: ?FATE_MAP_T.
+-type fate_string()  :: ?FATE_STRING_T.
+-type fate_address() :: ?FATE_ADDRESS_T.
+-type fate_variant(T):: ?FATE_VARIANT_T(T).
+
+-type fate_void()    :: ?FATE_VOID_T.
+
+-type fate_tuple(T) :: fate_unit()
+                     | ?FATE_TUPLE_T({T})
+                     | ?FATE_TUPLE_T({T,T})
+                     | ?FATE_TUPLE_T({T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T,T,T}).
+
+-type fate_none_map_type() ::
+        fate_boolean()
+      | fate_integer()
+      | fate_nil()
+      | fate_list(fate_none_map_type())
+      | fate_unit()
+      | fate_tuple(fate_none_map_type())
+      | fate_string()
+      | fate_address()
+      | fate_variant(fate_none_map_type()).
+
+-type fate_none_void_type() :: fate_none_map_type()
+                             | fate_map()
+                             | fate_list(fate_none_void_type())
+                             | fate_tuple(fate_none_void_type())
+                             | fate_variant(fate_none_void_type()).
+
+-type fate_type() :: fate_none_void_type() | fate_void().
+
+-export_type([fate_type/0]).
+-export_type([fate_none_void_type/0]).
+-export([ make_integer/1
+        , make_boolean/1
+        , make_list/1
+        , make_variant/2
+        , make_tuple/1
+        , make_string/1
+        , make_map/1
+        , make_address/1
+        , make_unit/0
+        ]).
+-export([format/1]).
+
+
+make_integer(I) when is_integer(I) -> ?MAKE_FATE_INTEGER(I).
+make_boolean(true)  -> ?FATE_TRUE;
+make_boolean(false) -> ?FATE_FALSE.
+make_list([]) -> ?FATE_NIL;
+make_list(L) -> ?MAKE_FATE_LIST(L).
+make_string(S) when is_list(S) -> ?FATE_STRING(list_to_binary(lists:flatten(S)));
+make_string(S) when is_binary(S) -> ?FATE_STRING(S).
+make_unit() -> ?FATE_UNIT.
+make_tuple(T) -> ?FATE_TUPLE(T).
+make_map(M) -> ?MAKE_FATE_MAP(M).
+make_address(A) -> ?FATE_ADDRESS(A).
+
+make_variant(Tag, Values) when is_integer(Tag)
+                               , 0 =< Tag
+                               , Tag < 256
+                               , is_tuple(Values) ->
+    ?FATE_VARIANT(Tag, Values).
+
+-spec format(fate_type()) -> iolist().
+format(I) when ?IS_FATE_INTEGER(I) -> integer_to_list(?MAKE_FATE_INTEGER(I));
+format(?FATE_VOID) -> "void";
+format(?FATE_TRUE) -> "true";
+format(?FATE_FALSE) -> "false";
+format(?FATE_NIL) -> "[]";
+format(L) when ?IS_FATE_LIST(L) ->
+    case ?FATE_LIST_VALUE(L) of
+        [E] -> "[ " ++ format(E) ++ " ]";
+        _ -> "[ " ++ format_list(L)
```

Comment:
`L` should be `?FATE_LIST_VALUE(L)`. And you don't need the special case for the singleton list.

Reply:
Fixed.